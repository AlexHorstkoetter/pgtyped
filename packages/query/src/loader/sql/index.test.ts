import parse from './index';

test('Named query', () => {
  const text = `
  /* @name GetAllUsers */
  SELECT * FROM users;`;
  const parseTree = parse(text);
  expect(parseTree).toMatchSnapshot();
});

test('Named query selects some fields', () => {
  const text = `
  /* @name GetAllUsers */
  SELECT id, name FROM users;`;
  const parseTree = parse(text);
  expect(parseTree).toMatchSnapshot();
});

test('Named query with an inferred param', () => {
  const text = `
  /* @name GetUserById */
  SELECT * FROM users WHERE userId = :userId;`;
  const parseTree = parse(text);
  expect(parseTree).toMatchSnapshot();
});

test('Named query with a valid param', () => {
  const text = `
  /*
    @name CreateCustomer
    @param customer -> (customerName, contactName, address)
  */
  INSERT INTO customers (customer_name, contact_name, address)
  VALUES :customers;`;
  const parseTree = parse(text);
  expect(parseTree).toMatchSnapshot();
});

test('Unused parameters produce warnings', () => {
  const text = `
  /*
    @name GetAllUsers
    @param userNames -> (...)
    @param users -> ((name,time)...)
  */
  SELECT * FROM users;`;
  const parseTree = parse(text);
  expect(parseTree).toMatchSnapshot();
});

test('Another test', () => {
  const text = `
    /* @name GetBooksByAuthorName */
    SELECT b.* FROM books b
    INNER JOIN authors a ON a.id = b.author_id
    WHERE a.first_name || ' ' || a.last_name = :authorName;`;
  const parseTree = parse(text);
  expect(parseTree).toMatchSnapshot();
});
