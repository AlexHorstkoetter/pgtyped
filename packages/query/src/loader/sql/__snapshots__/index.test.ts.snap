// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Named query 1`] = `
Object {
  "events": Array [],
  "parseTree": Object {
    "queries": Array [
      Object {
        "name": "GetAllUsers",
        "params": Array [],
        "statement": Object {
          "body": "SELECT * FROM users",
          "loc": Object {
            "a": 29,
            "b": 47,
            "col": 2,
            "line": 3,
          },
        },
        "usedParamSet": Set {},
      },
    ],
  },
}
`;

exports[`Named query selects some fields 1`] = `
Object {
  "events": Array [
    Object {
      "location": Object {
        "a": undefined,
        "b": undefined,
        "col": 11,
        "line": 3,
      },
      "message": Object {
        "text": "Parse error: token recognition error at: ','",
        "type": 0,
      },
      "type": 2,
    },
  ],
  "parseTree": Object {
    "queries": Array [
      Object {
        "name": "GetAllUsers",
        "params": Array [],
        "statement": Object {
          "body": "SELECT id, name FROM users",
          "loc": Object {
            "a": 29,
            "b": 54,
            "col": 2,
            "line": 3,
          },
        },
        "usedParamSet": Set {},
      },
    ],
  },
}
`;

exports[`Named query with a valid param 1`] = `
Object {
  "events": Array [
    Object {
      "location": Object {
        "a": undefined,
        "b": undefined,
        "col": 24,
        "line": 6,
      },
      "message": Object {
        "text": "Parse error: token recognition error at: '('",
        "type": 0,
      },
      "type": 2,
    },
    Object {
      "location": Object {
        "a": undefined,
        "b": undefined,
        "col": 38,
        "line": 6,
      },
      "message": Object {
        "text": "Parse error: token recognition error at: ','",
        "type": 0,
      },
      "type": 2,
    },
    Object {
      "location": Object {
        "a": undefined,
        "b": undefined,
        "col": 52,
        "line": 6,
      },
      "message": Object {
        "text": "Parse error: token recognition error at: ','",
        "type": 0,
      },
      "type": 2,
    },
    Object {
      "location": Object {
        "a": undefined,
        "b": undefined,
        "col": 61,
        "line": 6,
      },
      "message": Object {
        "text": "Parse error: token recognition error at: ')'",
        "type": 0,
      },
      "type": 2,
    },
    Object {
      "location": Object {
        "a": 42,
        "b": 49,
        "col": 11,
        "line": 4,
      },
      "message": Object {
        "text": "Parameter \\"customer\\" is defined but never used",
        "type": 0,
      },
      "type": 1,
    },
  ],
  "parseTree": Object {
    "queries": Array [
      Object {
        "name": "CreateCustomer",
        "params": Array [
          Object {
            "codeRefs": Object {
              "defined": Object {
                "a": 42,
                "b": 49,
                "col": 11,
                "line": 4,
              },
            },
            "name": "customer",
            "transform": Object {
              "keys": Array [
                "customerName",
                "contactName",
                "address",
              ],
              "type": "pick_tuple",
            },
          },
          Object {
            "codeRefs": Object {
              "used": Object {
                "a": 169,
                "b": 177,
                "col": 10,
                "line": 7,
              },
            },
            "name": "customers",
            "transform": Object {
              "type": "scalar",
            },
          },
        ],
        "statement": Object {
          "body": "INSERT INTO customers (customer_name, contact_name, address)
  VALUES :customers",
          "loc": Object {
            "a": 98,
            "b": 177,
            "col": 2,
            "line": 6,
          },
        },
        "usedParamSet": Set {
          "customers",
        },
      },
    ],
  },
}
`;

exports[`Named query with an inferred param 1`] = `
Object {
  "events": Array [],
  "parseTree": Object {
    "queries": Array [
      Object {
        "name": "GetUserById",
        "params": Array [
          Object {
            "codeRefs": Object {
              "used": Object {
                "a": 65,
                "b": 70,
                "col": 38,
                "line": 3,
              },
            },
            "name": "userId",
            "transform": Object {
              "type": "scalar",
            },
          },
        ],
        "statement": Object {
          "body": "SELECT * FROM users WHERE userId = :userId",
          "loc": Object {
            "a": 29,
            "b": 70,
            "col": 2,
            "line": 3,
          },
        },
        "usedParamSet": Set {
          "userId",
        },
      },
    ],
  },
}
`;

exports[`Unused parameters produce warnings 1`] = `
Object {
  "events": Array [
    Object {
      "location": Object {
        "a": 39,
        "b": 47,
        "col": 11,
        "line": 4,
      },
      "message": Object {
        "text": "Parameter \\"userNames\\" is defined but never used",
        "type": 0,
      },
      "type": 1,
    },
    Object {
      "location": Object {
        "a": 69,
        "b": 73,
        "col": 11,
        "line": 5,
      },
      "message": Object {
        "text": "Parameter \\"users\\" is defined but never used",
        "type": 0,
      },
      "type": 1,
    },
  ],
  "parseTree": Object {
    "queries": Array [
      Object {
        "name": "GetAllUsers",
        "params": Array [
          Object {
            "codeRefs": Object {
              "defined": Object {
                "a": 39,
                "b": 47,
                "col": 11,
                "line": 4,
              },
            },
            "name": "userNames",
            "transform": Object {
              "type": "array_spread",
            },
          },
          Object {
            "codeRefs": Object {
              "defined": Object {
                "a": 69,
                "b": 73,
                "col": 11,
                "line": 5,
              },
            },
            "name": "users",
            "transform": Object {
              "keys": Array [
                "name",
                "time",
              ],
              "type": "pick_array_spread",
            },
          },
        ],
        "statement": Object {
          "body": "SELECT * FROM users",
          "loc": Object {
            "a": 102,
            "b": 120,
            "col": 2,
            "line": 7,
          },
        },
        "usedParamSet": Set {},
      },
    ],
  },
}
`;
